## Data Fields Exploration
### *This notebook explores the data fields in the USGS Wind Turbine Data Fields*

In [1]:
link_to_metadata = 'https://eerscmap.usgs.gov/uswtdb/assets/data/uswtdb_v1_1_20180710.xml'
link_to_usgs_data = 'https://eerscmap.usgs.gov/uswtdb/assets/data/uswtdbCSV.zip'

In [2]:
import re
import sys
import xml.etree.ElementTree as ET
from urllib.request import urlopen
import pandas as pd
from bs4 import BeautifulSoup as BS

In [3]:
usgs_data = pd.read_csv('./uswtdbCSV/uswtdb_v1_1_20180710.csv')

In [4]:
usgs_data.shape

(57646, 24)

* There are 57646 rows in the dataset with 24 columns.

In [5]:
usgs_data.columns

Index(['case_id', 'faa_ors', 'faa_asn', 'usgs_pr_id', 't_state', 't_county',
       't_fips', 'p_name', 'p_year', 'p_tnum', 'p_cap', 't_manu', 't_model',
       't_cap', 't_hh', 't_rd', 't_rsa', 't_ttlh', 't_conf_atr', 't_conf_loc',
       't_img_date', 't_img_srce', 'xlong', 'ylat'],
      dtype='object')

In [8]:
print(usgs_data.head().to_string())

   case_id  faa_ors  faa_asn  usgs_pr_id t_state     t_county  t_fips    p_name  p_year  p_tnum  p_cap  t_manu  t_model  t_cap    t_hh    t_rd   t_rsa  t_ttlh  t_conf_atr  t_conf_loc  t_img_date     t_img_srce       xlong       ylat
0  3073381  missing  missing        5777      CA  Kern County    6029  251 Wind    1987     194  18.43  Vestas  missing     95 -9999.0 -9999.0 -9999.0 -9999.0           2           3  11/23/2017  Digital Globe -118.353210  35.086182
1  3073379  missing  missing        5763      CA  Kern County    6029  251 Wind    1987     194  18.43  Vestas  missing     95 -9999.0 -9999.0 -9999.0 -9999.0           2           3  11/23/2017  Digital Globe -118.354362  35.085335
2  3003862  missing  missing        5836      CA  Kern County    6029  251 Wind    1987     194  18.43  Vestas  missing     95 -9999.0 -9999.0 -9999.0 -9999.0           2           3   7/31/2017  Digital Globe -118.351288  35.091396
3  3001889  missing  missing        5110      CA  Kern County    602

* To explore the meaning of the columns in the dataset, we will need to traverse the XML Tree of the metadata.
* I downloaded the raw XML file from the website and saved it in my working directory from [HERE](https://eerscmap.usgs.gov/uswtdb/assets/data/uswtdb_v1_1_20180710.xml).

In [14]:
tree = ET.parse('./uswtdb_v1_1_20180710.xml')
root = tree.getroot()

In [21]:
ean_info = root.find('eainfo')
all_attributes = ean_info.find('detailed')

In [55]:
counter = 0
for elem in all_attributes:
    if elem.tag != 'attr':
        continue
    counter += 1
    try:
        attr_label = elem.find('attrlabl').text
    except:
        attr_label = None
    try:
        attr_def = elem.find('attrdef').text
    except:
        attr_def = None
    try:
        attr_defs = elem.find('attrdefs').text
    except:
        attr_defs = None
    try:
        attr_domv = elem.find('attrdomv').find('udom').text
    except:
        try:
            attr_domv = elem.find('attrdomv').find('rdom').find('attrunit').text
        except:
            attr_domv = None
    print('#', counter, attr_label)
    print('\n')
    print('Definition: ')
    print(attr_def)
    print('\n')
    print('Additional Details: ')
    print(attr_defs)
    print('\n')
    print('Attribute Units Explanation: ')
    print(attr_domv)
    print('- - - - - - - - -')

# 1 case_id


Definition: 
unique stable identification number


Additional Details: 
Producer defined


Attribute Units Explanation: 
1
- - - - - - - - -
# 2 faa_ors


Definition: 
faa unique identifier for each turbine for cross-reference to the faa digital obstacle files (faa dof )


Additional Details: 
FAA, https://www.faa.gov/air_traffic/flight_info/aeronav/digital_products/dof/


Attribute Units Explanation: 
identifier with the first two digits indicating a state, then a dash, with a sequential number following; "missing" records are ones with no known DOF number
- - - - - - - - -
# 3 faa_asn


Definition: 
faa obstruction evaluation airport airspace analysis (oe-aaa) aeronautical study number (asn)


Additional Details: 
FAA, https://oeaaa.faa.gov/oeaaa/external/public/publicAction.jsp


Attribute Units Explanation: 
year-region-number-case type (faa regional boundaries id across the country such as: AAL, ACE, AEA, AGL, ANE, ANM, ASO, ASW, AWP, WTE and WTW) (case types: Nonrul